File: 144.二叉树的前序遍历.java
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

import datastruct.TreeNode;

/*
 * @lc app=leetcode.cn id=144 lang=java
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }

        List<Integer> res = new ArrayList<>();
        Deque<TreeNode> stack = new LinkedList<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode p = stack.pop();
            res.add(p.val);

            if (p.right != null) {
                stack.push(p.right);
            }
            if (p.left != null) {
                stack.push(p.left);
            }
        }
        return res;
    }
}
// @lc code=end

File: 144.二叉树的前序遍历.kt
/*
 * @lc app=leetcode.cn id=144 lang=kotlin
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun preorderTraversal(root: TreeNode?): List<Int> {
        val outputList = mutableListOf<Int>()

        if (root == null) {
            return outputList
        }

        val stack = mutableListOf<TreeNode>()
        stack.add(root)

        while (stack.isNotEmpty()) {
            val p = stack.removeAt(stack.lastIndex)
            outputList.add(p.`val`)
            if (p.right != null) {
                stack.add(p.right!!)
            }
            if (p.left != null) {
                stack.add(p.left!!)
            }
        }

        return outputList
    }
}
// @lc code=end

